File: agent/api-client.ts
export type Category = {
  id: number;
  name: string;
  slug: string;
  seeded: boolean;
};

export type Resource = {
  id: number;
  url: string;
  title: string;
  description: string | null;
  category: Category;
};

type ApiEnvelope<T> = {
  success: boolean;
  data?: T;
  error?: unknown;
};

function requireEnv(name: string): string {
  const value = process.env[name]?.trim();
  if (!value) {
    throw new Error(`${name} is required`);
  }
  return value;
}

function boardApiUrl(): string {
  return requireEnv("BOARD_API_URL").replace(/\/+$/, "");
}

function jsonHeaders(): HeadersInit {
  return {
    "Content-Type": "application/json",
    "x-api-key": requireEnv("BOARD_API_SECRET"),
  };
}

async function parseJson<T>(response: Response): Promise<T> {
  return response.json() as Promise<T>;
}

export async function postResource(
  url: string,
  title: string,
  description: string | null,
  categorySlug: string,
): Promise<ApiEnvelope<Resource>> {
  const response = await fetch(`${boardApiUrl()}/api/resources`, {
    method: "POST",
    headers: jsonHeaders(),
    body: JSON.stringify({ url, title, description, categorySlug }),
    signal: AbortSignal.timeout(15_000),
  });
  return parseJson<ApiEnvelope<Resource>>(response);
}

export async function getResources(category?: string, limit = 10): Promise<Resource[]> {
  const params = new URLSearchParams({ limit: String(limit) });
  if (category) {
    params.set("category", category);
  }

  const response = await fetch(`${boardApiUrl()}/api/resources?${params.toString()}`, {
    signal: AbortSignal.timeout(10_000),
  });
  if (!response.ok) {
    throw new Error(`Failed to fetch resources: ${response.status}`);
  }

  const payload = await parseJson<ApiEnvelope<Resource[]>>(response);
  return payload.data ?? [];
}

export async function deleteResource(resourceId: number): Promise<boolean> {
  const response = await fetch(`${boardApiUrl()}/api/resources/${resourceId}`, {
    method: "DELETE",
    headers: jsonHeaders(),
    signal: AbortSignal.timeout(10_000),
  });
  return response.ok;
}

export async function searchResources(query: string): Promise<Resource[]> {
  const params = new URLSearchParams({ q: query, limit: "10" });
  const response = await fetch(`${boardApiUrl()}/api/resources?${params.toString()}`, {
    signal: AbortSignal.timeout(10_000),
  });
  if (!response.ok) {
    throw new Error(`Failed to search resources: ${response.status}`);
  }

  const payload = await parseJson<ApiEnvelope<Resource[]>>(response);
  return payload.data ?? [];
}

export async function getCategories(): Promise<Category[]> {
  const response = await fetch(`${boardApiUrl()}/api/categories`, {
    signal: AbortSignal.timeout(10_000),
  });
  if (!response.ok) {
    throw new Error(`Failed to fetch categories: ${response.status}`);
  }

  const payload = await parseJson<ApiEnvelope<Category[]>>(response);
  return payload.data ?? [];
}

export async function addCategory(name: string, slug: string): Promise<ApiEnvelope<Category>> {
  const response = await fetch(`${boardApiUrl()}/api/categories`, {
    method: "POST",
    headers: jsonHeaders(),
    body: JSON.stringify({ name, slug }),
    signal: AbortSignal.timeout(10_000),
  });
  return parseJson<ApiEnvelope<Category>>(response);
}

File: agent/api_client.py
"""HTTP client for the Curator Board API."""
import os
import httpx

BOARD_API_URL = os.environ["BOARD_API_URL"].rstrip("/")
BOARD_API_SECRET = os.environ["BOARD_API_SECRET"]

HEADERS = {
    "Content-Type": "application/json",
    "x-api-key": BOARD_API_SECRET,
}


async def post_resource(url: str, title: str, description: str | None, category_slug: str) -> dict:
    async with httpx.AsyncClient(timeout=15) as client:
        res = await client.post(
            f"{BOARD_API_URL}/api/resources",
            json={"url": url, "title": title, "description": description, "categorySlug": category_slug},
            headers=HEADERS,
        )
        res.raise_for_status()
        return res.json()


async def get_resources(category: str | None = None, limit: int = 10) -> list[dict]:
    params: dict = {"limit": limit}
    if category:
        params["category"] = category
    async with httpx.AsyncClient(timeout=10) as client:
        res = await client.get(f"{BOARD_API_URL}/api/resources", params=params)
        res.raise_for_status()
        return res.json().get("data", [])


async def delete_resource(resource_id: int) -> bool:
    async with httpx.AsyncClient(timeout=10) as client:
        res = await client.delete(
            f"{BOARD_API_URL}/api/resources/{resource_id}",
            headers=HEADERS,
        )
        return res.status_code == 200


async def search_resources(query: str) -> list[dict]:
    async with httpx.AsyncClient(timeout=10) as client:
        res = await client.get(
            f"{BOARD_API_URL}/api/resources",
            params={"q": query, "limit": 10},
        )
        res.raise_for_status()
        return res.json().get("data", [])


async def get_categories() -> list[dict]:
    async with httpx.AsyncClient(timeout=10) as client:
        res = await client.get(f"{BOARD_API_URL}/api/categories")
        res.raise_for_status()
        return res.json().get("data", [])


async def add_category(name: str, slug: str) -> dict:
    async with httpx.AsyncClient(timeout=10) as client:
        res = await client.post(
            f"{BOARD_API_URL}/api/categories",
            json={"name": name, "slug": slug},
            headers=HEADERS,
        )
        res.raise_for_status()
        return res.json()
